### PR TITLE
liquidity ads

### PR DESCRIPTION
**Pull request by niftynei**
Implements lightningnetwork/lightning-rfc#878

### CODE REVIEW

File: bitcoin/script.c [hunk 1]
```diff
@@ -521,15 +529,35 @@ u8 **bitcoin_witness_sig_and_element(const tal_t *ctx,
...
+ *      <revocationpubkey>
+ *  OP_ELSE
+ *      MAX(`to_self_delay`, `lease_end` - `blockheight`)
+ *      OP_CHECKSEQUENCEVERIFY
+ *      OP_DROP
+ *      <local_delayedpubkey>
+ *  OP_ENDIF
+ *  OP_CHECKSIG
+ */
+u8 *bitcoin_wscript_to_local(const tal_t *ctx, u16 to_self_delay, u32 csv,
```
Comment (rustyrussell) [inline bitcoin/script.c:552]: csv is a terrible name for this parameter!  lease_remaining?

File: doc/lightning-parsefeerate.7.md [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+--------
+
+**parsefeerate** *feerate_str*
+
+DESCRIPTION
+-----------
+
+The **parsefeerate** command returns the current feerate for any valid
+*feerate_str*. This is useful for finding the current feerate that a
+**fundpsbt** or **utxopsbt** command might use.
```
Comment (rustyrussell) [inline doc/lightning-parsefeerate.7.md:14]: This whole commit seems unnecessary?  `feerates` does this already?

**Review by rustyrussell — APPROVED** [review]
Ack e364d06c24123cd4eeb7297cde6dad1bab5f987c

File: gossipd/gossipd.c [hunk 3]
```diff
@@ -1418,8 +1419,11 @@ static struct io_plan *handle_new_lease_rates(struct io_conn *conn,
 	if (!fromwire_gossipd_new_lease_rates(msg, rates))
 		master_badmsg(WIRE_GOSSIPD_NEW_LEASE_RATES, msg);
 
-	tal_free(daemon->rates);
-	daemon->rates = rates;
+	daemon->rates = tal_free(daemon->rates);
+	if (!lease_rates_empty(rates))
+		daemon->rates = rates;
```
[the remark this answers is not in the document]
Comment (rustyrussell) [reply gossipd/gossipd.c:1424]: Np, fixed later.

**Comment by rustyrussell** [comment]
Added some neatening commits on the end, and trivial rebase now first commit is merged.

File: openingd/dualopend_wire.csv [hunk 4]
```diff
@@ -221,3 +222,14 @@ msgdata,dualopend_dry_run,channel_id,channel_id,
...
+# dualopend -> master:  validate liqudity offer sig
+msgtype,dualopend_validate_lease,7027
+msgdata,dualopend_validate_lease,sig,secp256k1_ecdsa_signature,
+msgdata,dualopend_validate_lease,lease_expiry,u32,
+msgdata,dualopend_validate_lease,chan_fee_max_base_msat,u32,
+msgdata,dualopend_validate_lease,chan_fee_max_ppt,u16,
+msgdata,dualopend_validate_lease,their_pubkey,pubkey,
+
+msgtype,dualopend_validate_lease_reply,7127
+msgdata,dualopend_validate_lease_reply,err_msg,?wirestring,
```
[the remark this answers is not in the document]
Comment (niftynei) [reply openingd/dualopend_wire.csv:249]: it seems made the (incorrect) assumption that only lightningd could verify these. Ooof. Will leave as is for now.

File: bitcoin/script.c (hunk 1 above)
Comment (niftynei) [reply bitcoin/script.c]: updated!

File: doc/lightning-parsefeerate.7.md (hunk 2 above)
Comment (niftynei) [reply doc/lightning-parsefeerate.7.md:14]: Not exactly, `feerates` returns the named current feerates; this takes any arbitrary feerate string and returns its value in perkw. I mean, technically, I could halfway parse the feerate string, figure out if it's a plain number or a named fee rate and if it's the latter match it up with the returned fields from `feerates`. Except that's a bit more work than just piping it through to lightningd to do it for me instead?

File: plugins/libplugin.c [hunk 5]
```diff
@@ -484,38 +484,72 @@ static struct json_out *start_json_request(const tal_t *ctx,
 }
 
 /* Synchronous routine to send command and extract fields from response */
-void rpc_scan(struct plugin *plugin,
-	      const char *method,
-	      const struct json_out *params TAKES,
-	      const char *guide,
-	      ...)
+static const char *rpc_scan_list(const tal_t *ctx,
```
Comment (niftynei) [inline plugins/libplugin.c]: I don't actually end up using this commit, I'm throwing it away.

**Comment by niftynei** [comment]
ACK https://github.com/ElementsProject/lightning/pull/4639/commits/7be8e71ca2b995313a8eba3156bf625a469fd762

**Pull request merged**